app/services/tsheets_service.rb

```diff
@@ -0,0 +1,29 @@
+class TsheetsService
+  def initialize
+    @adapter = TsheetsAdapter.new
+  end
+
+  def sync_users
+    unable_to_sync, synced = [], []
+    @tsheets_users = @adapter.find_users
+    @tsheets_users.each do |user|
```

Comment:
This looks like a very database heavy operation, Sam. From reading your code i understood that you take every user and do these calls:
- Check that tsheet_user_id doesn't exist
- Check users with similar names
- Update with tsheet_user_id.

I would try to do all this in batches, you can easily check all tsheet_user.id that are not in database with one database call. You can also easily update in batches (I'll attach example code later).